neural_lam/datastore/base.py

```diff
@@ -0,0 +1,371 @@
+        space and time) of a given category (state/forcin g/static). A
+        datastore must be able to return for the "state" category, but
+        "forcing" and "static" are optional (in which case the method should
+        return `None`).
+
+        The returned dataarray is expected to at minimum have dimensions of
+        `(grid_index, {category}_feature)` so that any spatial dimensions have
+        been stacked into a single dimension and all variables and levels have
+        been stacked into a single feature dimension named by the `category` of
+        data being loaded.
+
+        For categories of data that have a time dimension (i.e. not static
+        data), the dataarray is expected additionally have `(analysis_time,
+        elapsed_forecast_duration)` dimensions if `is_forecast` is True, or
+        `(time)` if `is_forecast` is False.
+
+        If the data is ensemble data, the dataarray is expected to have an
+        additional `ensemble_member` dimension.
+
+        Parameters
+        ----------
+        category : str
+            The category of the dataset (state/forcing/static).
+        split : str
+            The time split to filter the dataset (train/val/test).
+
+        Returns
+        -------
+        xr.DataArray or None
+            The xarray DataArray object with processed dataset.
+
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def boundary_mask(self) -> xr.DataArray:
+        """
+        Return the boundary mask for the dataset, with spatial dimensions
+        stacked. Where the value is 1, the grid point is a boundary point, and
+        where the value is 0, the grid point is not a boundary point.
+
+        Returns
+        -------
+        xr.DataArray
+            The boundary mask for the dataset, with dimensions
+            `('grid_index',)`.
+
+        """
+        pass
+
+
+@dataclasses.dataclass
+class CartesianGridShape:
+    """Dataclass to store the shape of a grid."""
+
+    x: int
+    y: int
+
+
+class BaseCartesianDatastore(BaseDatastore):
+    """Base class for weather
+    data stored on a Cartesian
+    grid. In addition to the
+    methods and attributes
+    required for weather data
+    in general (see
+    `BaseDatastore`) for
+    Cartesian gridded source
+    data each `grid_index`
+    coordinate value is assume
+    to have an associated `x`
+    and `y`-value so that the
+    processed data-arrays can
+    be reshaped back into into
+    2D xy-gridded arrays.
+
+    In addition the following attributes and methods are required:
+    - `coords_projection` (property): Projection object for the coordinates.
+    - `grid_shape_state` (property): Shape of the grid for the state variables.
+    - `get_xy_extent` (method): Return the extent of the x, y coordinates for a
+      given category of data.
+    - `get_xy` (method): Return the x, y coordinates of the dataset.
+
+    """
```

Comment:
I think having this subclassing to `BaseCartesianDatastore` is good, but I wonder if everything that relates to coordinates should be delegated here. Even if a datastore provides data that is not on a rectangular cartesian grid it will still have some coordinates and a projection.

In particular I am thinking that 
1. the `coords_projection` property could belong to the `BaseDataStore`
2. The `stacked=True` version of `get_xy` could also belong to `BaseDataStore` (then maybe better called `get_coords`). This would of course complicate having the two versions (`stacked=True/False`) of `get_xy`, but perhaps it would suffice to have the stacked version? Since `unstack_grid_coords` exists you can always use `grid_coords = cds.unstack_grid_coords(cds.get_xy("state"))` to get a non-stacked version.

Why I started thinking about this is because we are moving to data setups that won't be on nice rectangular grids, and will have to change up how e.g. plotting is done accordingly. If we move things related to coordinates in the most general sense to the most general base class we avoid having to implement much duplicate things in a `BaseNonCartesianDataStore` class  (not the name I would use, but you get the idea :laughing:). But I might be missing something here that makes this hard or impractical, so please let me know.

Reply:
I agree about putting the projection information in the base class, but the way I am thinking of stacking this an operation that only applies from going from 2D x/y arrays to flattining the spatial coordinates in 1D arrays containing the x/y coordinate values. True, we should have a `BaseDataStore.get_xy()` function, but maybe it doesn't make sense for it to have a `stacked` argument since it _has to_ be "stacked". Although this would mean the implementations of `BaseCartesianDatastore` and `BaseDatastore` would have a different function signature I think that makes sense. My expectation is that any function that relies on using "unstacked" arrays (i.e. plotting that uses `pcolormesh`) would check that the provided datastore is an implementation of `BaseCartesianDatastore` and if not will either raise an exception or try a different plotting approach that works for data not on a regular grid.

In general "cartesian" is maybe the wrong name here. What I am trying to distinguish is whether the source data is on a regular or irregular grid. Maybe I should rename this `RegularGridBaseDatastore`? Since "cartesian" to me more suggests that the grid has x/y-coordinates, which all our data is assumed to have (for now at least)

Reply:
I hope my above comment makes sense. For now I was trying to laying the future structure that will help in working with irregular and regular gridded data. I was also thinking this would be needed if new architectures were added that needs regular gridded data.

I would say though that we could go with the implementation I have and iterate on this as we realise more functionality we would like

Reply:
> True, we should have a BaseDataStore.get_xy() function, but maybe it doesn't make sense for it to have a stacked argument since it has to be "stacked".

Yes, fully agree. The concept of "stacked" should not exist in the `BaseDataStore`.

> Although this would mean the implementations of BaseCartesianDatastore and BaseDatastore would have a different function signature I think that makes sense.

You could also just introduce a `get_unstacked_xy` as a separate function in `BaseCartesianDatastore` to avoid such confusion.

> For now I was trying to laying the future structure that will help in working with irregular and regular gridded data. I was also thinking this would be needed if new architectures were added that needs regular gridded data.

I can see the value in this and I think the structure makes sense. The only things I propose to change are the exact points 1 and 2, otherwise I think this is all good.

Reply:
Leif already implemented this, going for the different function signatures. I added a comment to `BaseCartesianDatastore.get_xy` that `stacked` is not present in the parent class.